---
layout: post
title: ä¸€æ¬¡Bowtie2+deepTools+MACS2çš„ChIP-seqåˆ†æè®°å½•
tags: [ChIP-seq, toolkits, pipline]
cover-img: /assets/img/covers/sanger_seq.png
thumbnail-img: /assets/img/mds/chip_profile.png
---

è›‹ç™½è´¨-DNAç›¸äº’ä½œç”¨å’Œè¡¨è§‚é—ä¼ æ ‡è®°çš„å…¨åŸºå› ç»„å›¾è°±å¯¹äºå…¨é¢ç†è§£è½¬å½•è°ƒæ§è¿‡ç¨‹ï¼Œç ´è¯‘æ„æˆå„ç§ç”Ÿç‰©è¿‡ç¨‹çš„åŸºå› è°ƒæ§ç½‘ç»œè‡³å…³é‡è¦ã€‚æ ¸å°ä½“å®šä½ä¸ DNA å’Œç»„è›‹ç™½çš„åŠ¨æ€ä¿®é¥°ç›¸ç»“åˆï¼Œåœ¨åŸºå› è°ƒæ§ä¸­å‘æŒ¥å…³é”®ä½œç”¨ï¼Œå¹¶ä¸”æŒ‡å¯¼å‘è‚²å’Œåˆ†åŒ–ã€‚æŸ“è‰²è´¨çŠ¶æ€çš„æ”¹å˜ï¼Œå¦‚é€šè¿‡å…è®¸æˆ–é˜»æ­¢ DNA ç»“åˆè›‹ç™½çš„æ¥è§¦ï¼Œæˆ–è€…ä¿®é¥°æ ¸å°ä½“è¡¨é¢ä»¥å¢å¼ºæˆ–é˜»ç¢æ•ˆåº”è›‹ç™½å¤åˆç‰©çš„å‹Ÿé›†ï¼Œèƒ½å¤Ÿç›´æ¥å½±å“è½¬å½•ã€‚æœ€è¿‘çš„ç ”ç©¶è¿›å±•è¡¨æ˜ï¼ŒæŸ“è‰²è´¨çŠ¶æ€å’Œè½¬å½•ä¹‹é—´çš„è¿™ç§ç›¸äº’ä½œç”¨æ˜¯åŠ¨æ€çš„ï¼Œå¹¶ä¸”æ¯”ä»¥å‰è®¤è¯†åˆ°çš„æ›´ä¸ºå¤æ‚ï¼Œå› æ­¤éœ€è¦å¯¹å¤šç§ç»†èƒç±»å‹å’Œé˜¶æ®µçš„è¡¨è§‚åŸºå› ç»„è¿›è¡Œç³»ç»Ÿåˆ†æï¼Œä»¥äº†è§£å‘è‚²è¿‡ç¨‹å’Œç–¾ç—…çŠ¶æ€ã€‚  
æŸ“è‰²è´¨å…ç–«æ²‰æ·€åæµ‹åº (ChIP-seq, Chromatin immunoprecipitation followed by sequencing) æ˜¯ä¸€ç§å¯¹ DNA ç»“åˆè›‹ç™½ã€ç»„è›‹ç™½ä¿®é¥°æˆ–æ ¸å°ä½“å®šä½è¿›è¡Œå…¨åŸºå› ç»„åˆ†æçš„æŠ€æœ¯ã€‚ç”±äºæ–°ä¸€ä»£æµ‹åºæŠ€æœ¯çš„å·¨å¤§è¿›æ­¥ï¼ŒChIP-seq æä¾›äº†æ›´é«˜çš„åˆ†è¾¨ç‡ã€æ›´ä½çš„å™ªéŸ³å’Œæ›´å¤§çš„è¦†ç›–èŒƒå›´ã€‚ChIP-seq å®éªŒäº§ç”Ÿå¤§é‡æ•°æ®ï¼Œæœ‰æ•ˆçš„è®¡ç®—åˆ†æå°†æå¤§çš„ä¿ƒè¿›ç”Ÿç‰©å­¦æœºåˆ¶çš„æ­ç¤ºã€‚

ChIP-seqçš„ä¸€èˆ¬åˆ†æè¿‡ç¨‹åŒ…æ‹¬ï¼š

- å»ºåº“æµ‹åºï¼Œè·å–readsï¼›
- å¯¹readsè¿›è¡Œè´¨æ§ï¼›
- å»é™¤æ¥å¤´å’Œä½è´¨é‡ç¢±åŸºï¼›
- å°†readsæ¯”å¯¹å›å‚è€ƒåŸºå› ç»„ï¼›
- ChIPè´¨é‡è¯„ä¼°ï¼›
- å»æ‰é‡å¤çš„readsï¼›
- æ£€æµ‹ä¿¡å·å¯Œé›†åŒºåŸŸï¼ˆpeakcallingï¼‰åŠæ³¨é‡Šï¼›
- IGVå¯è§†åŒ–peakï¼›
- Motifåˆ†æï¼›
- åŸºå› åŠŸèƒ½å¯Œé›†åˆ†æç­‰ï¼›
- å·®å¼‚åˆ†ææˆ–è€…è”åˆå¤šç»„å­¦åˆ†æã€‚

å¸¸ç”¨å·¥å…·æœ‰ï¼š

1. è´¨æ§ï¼š
[FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/)ã€[fastp](https://github.com/OpenGene/fastp)ã€[RSeQC](http://rseqc.sourceforge.net/)ç­‰
2. é¢„å¤„ç†readsï¼š
[Cutadapt](https://cutadapt.readthedocs.io/en/stable/)ã€[Trim Galore (Cutadapt+fastQC)](https://www.bioinformatics.babraham.ac.uk/projects/trim_galore/)ã€[Trimmomatic](http://www.usadellab.org/cms/?page=trimmomatic)ç­‰
3. readså›è´´ï¼š
[Bowtie2](http://bowtie-bio.sourceforge.net/bowtie2/index.shtml)ã€[BWA](http://bio-bwa.sourceforge.net/)
4. è¿‡æ»¤é‡å¤readsï¼š
[Picard](http://broadinstitute.github.io/picard/)ã€[Samtools](https://www.htslib.org/doc/samtools.html)
5. peakcallingå’Œæ³¨é‡Š:
[MACS](https://github.com/macs3-project/MACS)ã€[Homer](http://homer.ucsd.edu/homer/)ã€[ChIP](https://bioconductor.org/packages/release/bioc/vignettes/ChIPseeker/inst/doc/ChIPseeker.html)ç­‰
6. Motifåˆ†æï¼š
[MEME](https://meme-suite.org/meme/)ã€[Homer](http://homer.ucsd.edu/homer/)
7. åŠŸèƒ½å¯Œé›†åˆ†æï¼š
[GREAT](http://bejerano.stanford.edu/great/public/html/)ç­‰
8. å·®å¼‚åˆ†æï¼š
[DiffBind](https://bioconductor.org/packages/release/bioc/html/DiffBind.html)
9. å¯è§†åŒ–ï¼š
[deepTools](https://deeptools.readthedocs.io/en/develop/)ã€[IGV](https://igv.org/)ã€[bedtools](https://bedtools.readthedocs.io/en/latest/content/bedtools-suite.html)ã€[WWebLogo](https://weblogo.threeplusone.com/)ç­‰

### readsæ•°æ®å¤„ç†

åŸºäºSnakemakeå®šä¹‰çš„RNA-seqåˆ†ææµç¨‹ï¼Œä½¿ç”¨äº†Bowtie2+deepToolsçš„ç»„åˆå·¥å…·ï¼Œé’ˆå¯¹åŒç«¯æµ‹åºï¼ˆpair-endï¼‰ä½¿ç”¨ï¼š
```
###################
# Genome files #
genome_index="/home/user/genomes/bowtie2Index/mm10/DNA/genome"
genome_gtf="/home/user/genomes/mm10.refGene.gtf"
genome_gtf_gene="/home/user/genomes/mm10.refGene.transcript.gtf"
blacklist="/home/user/genomes/mm10_blacklist_ENCFF547MET.bed"
###################
# input files #
species="Mouse/Mus musculus"  #"Human/Homo sapiens"
reads=["_R1","_R2"]           #_1,_2
ext=".fastq.gz"               #.fq.gz
path_origin=""                #æµ‹åºåŸå§‹æ•°æ®å­˜æ”¾è·¯å¾„
sample_input=[]               #inputæ ·æœ¬ï¼Œä¸åŠ åç¼€
sample_chip=[]                #chipæ ·æœ¬
###################
# pre-pipeline #
import os
import sys
from pathlib import Path

samples=sample_input+sample_chip
p = Path(path_origin)
ss=list(p.glob("**/*"+ext))  ## æ³¨æ„ï¼Œ**ç›®å½•é“¾æ¥æ— æ³•æ‰¾åˆ°æ–‡ä»¶
l=len(ext)+len(reads[0])

Path("rawdata").mkdir(parents=True, exist_ok=True)
pwd=Path.cwd()
for s in ss:
    f=s.name
    if f[0:-l] in samples:
        s2=pwd / "rawdata"/ f
        #s2.unlink(missing_ok=True)  #python 3.8æ”¯æŒ
        if s2.exists():
            os.remove(s2)
        s2.symlink_to(s)  #s2ä¸ºç”Ÿæˆçš„è½¯é“¾æ¥æ–‡ä»¶
        #os.symlink(s,s2)

print("#"*10)
print("Species is:")
print("All samples:", samples)
print("#"*10)
###################
rule all:
    input:
        expand("fastQC/{sample}{read}_fastqc.html",sample=samples,read=reads),
        expand("alignment/{sample}.sorted.bam",sample=samples),
        expand("alignment/{sample}.sorted.bam.bai",sample=samples),
        #expand("chipQC/{sample}_qc.txt", sample=samples),
        expand("deeptools/bigwig/{sample}.filter.dedup.bw", sample=samples),
        expand("deeptools/fragsize/{sample}_fragsize.png", sample=samples),
        expand("deeptools/profile/{sample}_gene.bed",sample=samples),
        expand("deeptools/profile/{sample}_gene.png",sample=samples),
        "deeptools/plotfingerprint.png",
        "deeptools/multibamsummary.counts.txt",
        "deeptools/correlation.png",

rule trimming:
    input:
        r1="rawdata/{sample}"+reads[0]+ext,
        r2="rawdata/{sample}"+reads[1]+ext
    output:
        r1="trimdata/{sample}"+reads[0]+ext,
        r2="trimdata/{sample}"+reads[1]+ext
    params:
        tmp1="trimdata/{sample}"+reads[0]+"_val_1.fq.gz",
        tmp2="trimdata/{sample}"+reads[1]+"_val_2.fq.gz"
    threads: 4
    log:
        "trimdata/logs/{sample}.log"
    shell:
        "trim_galore --paired --stringency 3 -q 20 --length 20 -o trimdata --cores {threads} {input.r1} {input.r2} 2> {log}; "
        "mv {params.tmp1} {output.r1}; "
        "mv {params.tmp2} {output.r2}"

rule fastQC:
    input:
        "trimdata/{sample}{read}"+ext
    output:
        "fastQC/{sample}{read}_fastqc.html"
    threads: 12
    log:
        "fastQC/logs/{sample}{read}.log"
    shell:
        "fastqc -o fastQC -t {threads} -q {input} 2> {log}"

rule bowtie2:
    input:
        r1="trimdata/{sample}"+reads[0]+ext,
        r2="trimdata/{sample}"+reads[1]+ext
    output:
        bam="alignment/{sample}.sorted.bam",
        summary="alignment/{sample}.summary"
    threads: 12
    params:
        index=genome_index
    log:
        "alignment/logs/{sample}.align.log"
    shell:
        "bowtie2 -q -X 1000 -x {params.index} -1 {input.r1} -2 {input.r2} -p {threads} 2> {output.summary} "
        "| samtools view -Sbh - | samtools sort -@ {threads} -o {output.bam} 2> {log}"

rule samindex:
    input:
        "alignment/{sample}.sorted.bam"
    output:
        "alignment/{sample}.sorted.bam.bai"
    shell:
        "samtools index {input}"

rule chipQC:
    input:
        "alignment/{sample}.sorted.bam"
    output:
        "chipQC/{sample}_qc.txt"
    log:
        "chipQC/logs/{sample}_qc.log"
    shell:"""
        bedtools bamtobed -bedpe -i {input} \
        | awk 'BEGIN{{OFS="\\t"}} {{print $1,$2,$3,$6}}' \
        | grep -v 'chrM' | sort | uniq -c \
        | awk 'BEGIN{{mt=0;m0=0;m1=0;m2=0}} ($1==1){{m1=m1+1}} ($1==2){{m2=m2+1}} {{m0=m0+1}} {{mt=mt+$1}} END{{m1_m2=-1.0; if(m2>0) m1_m2=m1/m2; printf "%d\\t%d\\t%d\\t%d\\t%f\\t%f\\t%f\\n",mt,m0,m1,m2,m0/mt,m1/m0,m1_m2}}' > {output}
        """
#echo -e "TotalReadPairs\\tDistinctReadPairs\\tOneReadPair\\tTwoReadPairs\\tNRF\\tPBC1\\tPBC2" > {output};  #é¦–è¡Œæ’å…¥è¡¨å¤´
#sed -i '1i/TotalReadPairs\tDistinctReadPairs\tOneReadPair\tTwoReadPairs\tNRF\tPBC1\tPBC2' {output}  #æ•ˆæœåŒä¸Š

rule fragmentszie:
    input:
        "alignment/{sample}.sorted.bam"
    output:
        png="deeptools/fragsize/{sample}_fragsize.png",
        hist="deeptools/fragsize/{sample}_fragsize.hist.txt"
    threads: 12
    log:
        "deeptools/logs/{sample}.fragsize.log"
    shell:
        "bamPEFragmentSize -b {input} -hist {output.png} --outRawFragmentLengths {output.hist} --maxFragmentLength 1000"

rule filter:
    input:
        "alignment/{sample}.sorted.bam"
    output:
        "alignment/filter/{sample}.filter.bam"
    threads: 12
    log:
        "alignment/filter/logs/{sample}.filter.log"
    shell:
        "samtools view -b -F 1804 -f 2 -q 30 -@ {threads} {input} > {output} 2> {log}"

rule deduplicates:
    input:
        "alignment/filter/{sample}.filter.bam"
    output:
        bam="alignment/filter/{sample}.filter.dedup.bam",
        matrix="alignment/filter/{sample}.filter.dedup.txt"
    threads: 16
    log:
        "alignment/filter/logs/{sample}.dedup.log"
    shell:
        "picard MarkDuplicates INPUT={input} OUTPUT={output.bam} METRICS_FILE={output.matrix} "
        "REMOVE_DUPLICATES=true CREATE_INDEX=true VALIDATION_STRINGENCY=LENIENT 2> {log}"

rule bamcoverage:
    input:
        "alignment/filter/{sample}.filter.dedup.bam"
    output:
        "deeptools/bigwig/{sample}.filter.dedup.bw"
    params:
        black=blacklist
    threads: 12
    log:
        "deeptools/logs/{sample}.bamcoverage.log"
    shell:
        "bamCoverage -b {input} -o {output} -p {threads} "
        "--binSize 10 --smoothLength 30 --normalizeUsing RPKM --ignoreForNormalization chrX chrY chrM --extendReads 2> {log}"
# --blackListFileName {params.black}

rule profile:
    input:
        "deeptools/bigwig/{sample}.filter.dedup.bw"
    output:
        bed="deeptools/profile/{sample}_gene.bed",
        matrix="deeptools/profile/{sample}_gene.matrix.gz"
    params:
        ref=genome_gtf_gene
    threads: 12
    log:
        "deeptools/logs/{sample}.profile.log"
    shell:
        "computeMatrix reference-point --referencePoint TSS -b 2500 -a 2500 -R {params.ref} "
        "-S {input} -o {output.matrix} --outFileSortedRegions {output.bed} --skipZeros -p {threads} 2> {log}"

rule plotprofile:
    input:
        "deeptools/profile/{sample}_gene.matrix.gz"
    output:
        "deeptools/profile/{sample}_gene.png"
    shell:
        "plotHeatmap -m {input} -out {output} --dpi 360"

rule plotfingerprint:
    input:
        expand("alignment/filter/{sample}.filter.dedup.bam",sample=samples)
    output:
        matrix="deeptools/plotfingerprint.txt",
        png="deeptools/plotfingerprint.png"
    threads: 12
    log:
        "deeptools/logs/plotfingerprint.log"
    shell:
        "plotFingerprint -b {input} -o {output.png} --outQualityMetrics {output.matrix} "
        "-p {threads} --ignoreDuplicates --extendReads 2> {log}"

rule correlation:
    input:
        expand("alignment/filter/{sample}.filter.dedup.bam",sample=samples)
    output:
        counts="deeptools/multibamsummary.counts.txt",
        npz="deeptools/multibamsummary.counts.npz"
    threads:12
    params:
        black=blacklist
    log:
        "deeptools/logs/multibamsummary.log"
    shell:
        "multiBamSummary bins --bamfiles {input} --blackListFileName {params.black} --outRawCounts {output.counts} -o {output.npz} --extendReads --binSize 2000 -p {threads} 2> {log}; "

rule plotcorrelation:
    input:
        "deeptools/multibamsummary.counts.npz"
    output:
        "deeptools/correlation.png"
    shell:
        "plotCorrelation -in {input} -o {output} --corMethod pearson --whatToPlot scatterplot --skipZeros --removeOutliers --log1p"
```

### ä½¿ç”¨MACS2 callpeakä»¥åŠHomeræ³¨é‡Špeak

```
#callpeak
macs2 callpeak -f BAMPE -t CHIP.bam -c INPUT.bam --outdir PATH_TO_OUTDIR -n PRE_NAME -g mm|hs -p 0.01;  #sharp-peakï¼Œ-pä¿ç•™peakçš„é˜ˆå€¼
macs2 callpeak -f BAMPE -t CHIP.bam -c INPUT.bam --outdir PATH_TO_OUTDIR -n PRE_NAME -g mm|hs -p 1e-5 --broad --broad-cutoff 0.01  #broad-peakï¼Œ-pè®¾ç½®ä¸åŒçš„å€¼ï¼Œå°†ä¼šå½±å“broad-peakçš„pzå€¼

#è¿‡æ»¤peak
cat PRE_NAME.Peak |awk -F'\t' '{if ($8>5) print $0}' > PRE_NAME_p1e5.Peak

#åˆå¹¶peak
mergePeaks -venn <filename> <primary peak file> [additional peak/annotation files...] > outfile

#Homeræ³¨é‡Špeak
annotatePeaks.pl <peak/BED file> <genome> > <output file>

#Motifåˆ†æ
# DNA
findMotifsGenome.pl peakfile.bed /mnt/d/ubuntu/genomes/mouse/mm10.fa motif_homer -size 200 -mask -dumpFasta
# RNA
findMotifsGenome.pl peakfile.bed /mnt/d/ubuntu/genomes/mouse/mm10.fa motif_homer -rna
```

### é¥¼å›¾å±•ç¤ºHomeræ³¨é‡Šçš„peakåˆ†å¸ƒæƒ…å†µ

è¿™é‡Œè¦æ³¨æ„Homeræ³¨é‡Šçš„æ–‡ä»¶è¯´æ˜ï¼Œ  
1) The first determines the distance to the nearest TSS and assigns the peak to that gene.  
2) The second determines the genomic annotation of the region occupied by the center of the peak/region.  
ä¹Ÿå°±æ˜¯è¯´ï¼ŒDistance to TSSå’ŒAnnotation ä¸¤åˆ—çš„ä¿¡æ¯ä¸å®Œå…¨ä¸€è‡´ï¼Œåˆ†åˆ«å¯¹åº”ä»¥ä¸Šä¸¤æ­¥éª¤çš„å¯¹åº”åŸºå› ã€‚

```R
library(tidyverse)

# peak é¶åŸºå› çš„å½’å±ï¼ŒåŒ…æ‹¬Promoter <Distance to TSS (-2500~500 bp)>, Intron <Annotation>, Exon <Annotation>ï¼ŒIntergenic<Annotation>

## gene idè½¬æ¢
# BiocManager::install("rtracklayer")
library(rtracklayer)
library(ggsci)
library(scales)
# è¯»å–gffæ–‡ä»¶
gff <- readGFF() # gffæ–‡ä»¶ï¼Œå¦‚mm10.refGene.gtf
# head(gff)
id_map <- gff[, c("transcript_id", "gene_id", "gene_name")]
id_map <- unique(id_map) # gene idå¯¹ç…§è¡¨
# id è½¬æ¢
f <- "" # æ³¨é‡Šæ–‡ä»¶
data <- read_tsv(f)
data2 <- data %>%
  select(1:4, 8, 10, 12:17, 19) %>%
  separate(col = Annotation, into = c("feature", "annotation"), sep = " \\(") %>%
  separate(col = annotation, into = c("transcript_id", "xxx"), sep = "[,\\)]")
data2 <- merge(data2, id_map, by = "transcript_id", all.x = T)
data2 <- select(data2, 2:6, 1, 17, 8:15)
# colnames(data2)
#  [1] "PeakID"
#  [2] "Chr"
#  [3] "Start"
#  [4] "End"
#  [5] "feature"
#  [6] "transcript_id"
#  [7] "gene_name"
#  [8] "Distance to TSS"
# [9] "Entrez ID"
# [10] "Nearest Unigene"
# [11] "Nearest Refseq"
# [12] "Nearest Ensembl"
# [13] "Gene Name"
# [14] "Gene Alias"
# [15] "Gene Type"
# åŸºå› ç‰¹å¾å½’ç±»
feature2 <- as.character(data2$feature)
feature2[feature2 == "non-coding"] <- "Intergenic"
feature2[feature2 == "intron"] <- "Intron"
feature2[feature2 == "TTS"] <- "Exon"
feature2[feature2 == "3' UTR"] <- "Exon"
feature2[feature2 == "5' UTR"] <- "Exon"
feature2[feature2 == "exon"] <- "Exon"
feature2[feature2 == "promoter-TSS"] <- "Promoter"
feature2[feature2 == "promoter"] <- "Promoter"
# table(feature2)
data2$feature2 <- feature2
data2 <- data2[!is.na(feature2), ] # å»æ‰æ²¡æœ‰featureçš„peak
# å®šä¹‰promoteråŒºåŸŸå¤§å°ï¼Œå¦‚ -2500bp ~ 500bp
feature2 <- as.character(data2$feature2)
feature3 <- c()
dist <- as.numeric(as.character(data2$`Distance to TSS`))
for (i in 1:length(dist)) {
  if (dist[i] > (-2500) & dist[i] < 500 & (feature2[i] != "Promoter")) {
    feature3 <- c(feature3, "Promoter")
  } else {
    feature3 <- c(feature3, feature2[i])
  }
}
# table(feature3)
data2$feature3 <- feature3
write.table(data2, paste0(f, "2"), row.names = F, sep = "\t")
# é¥¼å›¾å±•ç¤ºHOMER æ³¨é‡Špeakçš„åˆ†å¸ƒæƒ…å†µ
mypal <- pal_npg("nrc", alpha = 0.6)(4)
show_col(mypal)
names(mypal) <- c("Intergenic", "Intron", "Exon", "Promoter")

data_pie <- data2 %>%
  group_by(feature3) %>%
  summarise(n = n()) %>%
  arrange(desc(n)) %>%
  mutate(prop = round(n / sum(n), 3) * 100) %>%
  mutate(ypos = cumsum(prop) - 0.5 * prop)
data_pie$feature3 <- factor(data_pie$feature3, levels = rev(data_pie$feature3))
data_pie$prop2 <- sprintf("%0.2f", data_pie$prop) ## ä¿ç•™å°æ•°åçš„0
data_pie <- na.omit(data_pie)
mypalx <- mypal[as.character(data_pie$feature3)]
data_pie %>%
  ggplot(aes(x = 2, y = prop, fill = feature3)) +
  geom_bar(stat = "identity", width = 1, color = "white") +
  coord_polar("y", start = 0) +
  # scale_fill_npg(alpha=0.8,)+
  scale_fill_manual(values = mypalx) +
  # geom_text(aes(label = paste0(prop, "%")), position = position_stack(vjust=0.5),colour='white')+
  geom_text(aes(y = ypos, label = paste0(prop2, "%")), color = "black", size = 4, nudge_x = 0.8) +
  theme_void() +
  labs(fill = "Features") +
  theme(text = element_text(size = 14), legend.key.height = unit(0.4, "cm"), legend.key.width = unit(0.4, "cm"))
ggsave(paste0(f, ".png"), width = 10, height = 10, units = "cm", dpi = 1200)
```

### IGV å¯è§†åŒ–peak

å¯è‡ªè¡Œç”Ÿæˆå‚è€ƒåŸºå› ç»„åŠç›¸åº”æ³¨é‡Šæ–‡ä»¶çš„genomeæ–‡ä»¶ã€‚

---

é“­è®°ä¹ä¸€å…«äº‹å˜Â·1931å¹´9æœˆ18æ—¥æ—¥æœ¬å…¨é¢ä¾µå ğŸ˜ ğŸ’¢ğŸ’¢


