---
layout: post
title: ä¸€æ¬¡Hisat2+featureCounts+DESeqçš„RNA-seqåˆ†æè®°å½•
tags: [RNA-seq, toolkits, pipline]
cover-img: /assets/img/covers/sanger_seq.png
thumbnail-img: /assets/img/mds/volcano_plot.png
---

åœ¨è¿‡å»çš„åå¹´ä¸­ï¼ŒRNA æµ‹åº (RNA-seq) æˆä¸ºäº†å…¨è½¬å½•ç»„åˆ†æå·®å¼‚åŸºå› è¡¨è¾¾å’ŒmRNAå·®å¼‚å‰ªæ¥çš„é‡è¦å·¥å…·ã€‚éšç€ä¸‹ä¸€ä»£æµ‹åºæŠ€æœ¯çš„å‘å±•ï¼ŒRNA-seqæ–¹æ³•å·²ç»è¢«å¹¿æ³›ç”¨äºç ”ç©¶RNAç”Ÿç‰©å­¦çš„è®¸å¤šä¸åŒé¢†åŸŸï¼ŒåŒ…æ‹¬å•ç»†èƒåŸºå› è¡¨è¾¾ã€ç¿»è¯‘å’ŒRNAç»“æ„ç­‰ã€‚å¦å¤–ï¼Œä¾‹å¦‚ç©ºé—´è½¬å½•ç»„å­¦ï¼ˆspatialomicsï¼‰ç­‰ä¸€äº›æ–°å…´çš„åº”ç”¨ä¹Ÿæ­£åœ¨æ¢ç´¢ä¹‹ä¸­ã€‚ç»“åˆæ–°çš„é•¿è¯»é•¿å’Œç›´æ¥RNA-seqæŠ€æœ¯ä»¥åŠç”¨äºæ•°æ®åˆ†æçš„æ›´å¥½çš„è®¡ç®—å·¥å…·çš„å¼€å‘ï¼ŒRNA-seqçš„åˆ›æ–°å°†æœ‰åŠ©äºæ›´å…¨é¢åœ°äº†è§£RNAç”Ÿç‰©å­¦ï¼ŒåŒ…æ‹¬è½¬å½•å‘ç”Ÿçš„æ—¶é—´å’Œä½ç½®ï¼Œä»¥åŠRNAåŠŸèƒ½è°ƒæ§é—®é¢˜ç­‰ã€‚

RNA-seqçš„å¸¸è§„åˆ†ææµç¨‹ä¸»è¦åŒ…æ‹¬ï¼š

- å»ºåº“æµ‹åºï¼Œè·å–readsï¼›
- å¯¹readsè¿›è¡Œè´¨æ§ï¼›
- å»é™¤æ¥å¤´å’Œä½è´¨é‡ç¢±åŸºï¼›
- å°†readsæ¯”å¯¹å›å‚è€ƒåŸºå› ç»„ï¼›
- ç»Ÿè®¡readsæ•°è¡¨å¾åŸºå› è¡¨è¾¾æ°´å¹³ï¼›
- åŸºå› å·®å¼‚è¡¨è¾¾åˆ†ææˆ–è€…è”åˆå¤šç»„å­¦åˆ†æï¼›
- åŸºå› åŠŸèƒ½å¯Œé›†åˆ†æç­‰ã€‚

å¸¸ç”¨å·¥å…·å¦‚ä¸‹ï¼š

1. è´¨æ§
[FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/)ã€[fastp](https://github.com/OpenGene/fastp)ã€[RSeQC](http://rseqc.sourceforge.net/)ç­‰
2. é¢„å¤„ç†reads
[Cutadapt](https://cutadapt.readthedocs.io/en/stable/)ã€[Trim Galore (Cutadapt+fastQC)](https://www.bioinformatics.babraham.ac.uk/projects/trim_galore/)ã€[Trimmomatic](http://www.usadellab.org/cms/?page=trimmomatic)ç­‰
3. readså›è´´
[HISAT2](http://daehwankimlab.github.io/hisat2/)ã€[STAR](https://github.com/alexdobin/STAR)
4. è®¡ç®—åŸºå› è¡¨è¾¾é‡
[featureCounts](http://subread.sourceforge.net/featureCounts.html)ã€[StringTie](https://ccb.jhu.edu/software/stringtie/)ã€[Salmon](https://salmon.readthedocs.io/en/latest/salmon.html)ã€[Cufflinks](https://github.com/cole-trapnell-lab/cufflinks)
Salmonå¯ä»¥é€‰æ‹©è·³è¿‡readsæ¯”å¯¹å›å‚è€ƒåŸºå› ç»„æ­¥éª¤ã€‚
5. å·®å¼‚åˆ†æ
[Cufflinks-cuffdiff](https://github.com/cole-trapnell-lab/cufflinks)ã€[DESeq2](https://bioconductor.org/packages/release/bioc/html/DESeq2.html)ã€[limma](https://bioconductor.org/packages/release/bioc/html/limma.html)ã€[edgeR](https://bioconductor.org/packages/release/bioc/html/edgeR.html)
6. åŠŸèƒ½å¯Œé›†åˆ†æ
[DAVID](https://david.ncifcrf.gov/home.jsp)ã€[GSEA](https://www.gsea-msigdb.org/gsea/index.jsp)ã€[Enrichr](https://maayanlab.cloud/Enrichr/)ã€[panther](http://pantherdb.org/)ã€[GREAT](http://bejerano.stanford.edu/great/public/html/)ã€[clusterProfiler](https://bioconductor.org/packages/release/bioc/html/clusterProfiler.html)ç­‰

[Snakemake](https://snakemake.readthedocs.io/en/stable/)æ˜¯åŸºäºPythonè¯­è¨€å®ç°çš„ä¸€ä¸ªå·¥ä½œæµç®¡ç†ç³»ç»Ÿï¼Œç”¨äºåˆ›å»ºå¯é‡å¤å’Œå¯æ‰©å±•çš„æ•°æ®åˆ†ææµç¨‹çš„å·¥å…·ã€‚Snakemake å¯ä»¥æ ¹æ®åˆ†ææ‰€éœ€è½¯ä»¶æè¿°ï¼Œè‡ªåŠ¨éƒ¨ç½²åˆ°ä»»ä½•æ‰§è¡Œç¯å¢ƒã€‚æ­¤å¤–ï¼Œå¯ä»¥æ— ç¼æ‰©å±•åˆ°æœåŠ¡å™¨ã€é›†ç¾¤ã€ç½‘æ ¼å’Œäº‘ç¯å¢ƒï¼Œæ— éœ€ä¿®æ”¹å·¥ä½œæµå®šä¹‰ã€‚

### readsæ•°æ®å¤„ç†

å¦‚ä¸‹ä¸ºåŸºäºSnakemakeå®šä¹‰çš„RNA-seqåˆ†ææµç¨‹ï¼Œä½¿ç”¨äº†Hisat2+featureCounts+DESeq2çš„ç»„åˆå·¥å…·ï¼Œé’ˆå¯¹åŒç«¯æµ‹åºï¼ˆpair-endï¼‰ä½¿ç”¨ï¼š
```
###################
# Genome files #
genome_index="/home/user/genomes/hisat2Index/hg38/genome"  #å‚è€ƒåŸºå› ç»„
genome_gtf="/home/user/genomes/hg38.ncbiRefSeq.gtf"        #å‚è€ƒåŸºå› ç»„æ³¨é‡Šæ–‡ä»¶
###################
# input files #
species="Human/Homo sapiens"  #
reads=["_R1","_R2"]           #æˆ–è€…_1,_2
ext=".fastq.gz"               #æˆ–è€….fq.gz
path_origin=""                #æµ‹åºåŸå§‹æ•°æ®å­˜æ”¾è·¯å¾„
samples=[]                    #æ ·æœ¬åï¼Œä¸åŠ åç¼€
###################
# pre-pipeline #
import os
import sys
from pathlib import Path


p = Path(path_origin)
ss=list(p.glob("**/*"+ext))   # æ³¨æ„ï¼Œ**ç›®å½•é“¾æ¥æ— æ³•æ‰¾åˆ°æ–‡ä»¶
l=len(ext)+len(reads[0])

Path("rawdata").mkdir(parents=True, exist_ok=True)
pwd=Path.cwd()
for s in ss:
    f=s.name
    if f[0:-l] in samples:
        s2=pwd / "rawdata"/ f
        #s2.unlink(missing_ok=True)  #python 3.8æ”¯æŒ
        if s2.exists():
            os.remove(s2)
        s2.symlink_to(s)  #æ³¨æ„s2ä¸ºç”Ÿæˆçš„è½¯é“¾æ¥æ–‡ä»¶
        #os.symlink(s,s2)

print("#"*10)
print("Species is:")
print("All samples:", samples)
print("#"*10)
###################
rule all:
    input:
        expand("fastQC/{sample}{read}_fastqc.html",sample=samples,read=reads),
        expand("alignment/{sample}.filter.sorted.bam.bai",sample=samples),
        expand("counts/{sample}.tab",sample=samples),
        expand("featurecounts/{sample}.count.txt",sample=samples),
        #"counts/gene_count_matrix.csv"

rule trimgalore:
    input:
        r1="rawdata/{sample}"+reads[0]+ext,
        r2="rawdata/{sample}"+reads[1]+ext
    output:
        r1="trimdata/{sample}"+reads[0]+ext,
        r2="trimdata/{sample}"+reads[1]+ext
    params:
        tmp1="trimdata/{sample}"+reads[0]+"_val_1.fq.gz",
        tmp2="trimdata/{sample}"+reads[1]+"_val_2.fq.gz"
    threads: 4
    log:
        "trimdata/logs/{sample}.log"
    shell:
        "trim_galore --paired --stringency 3 -q 20 --length 20 -o trimdata --cores {threads} {input.r1} {input.r2} 2> {log}; "
        "mv {params.tmp1} {output.r1}; "
        "mv {params.tmp2} {output.r2}"

rule fastqc:
    input:
        "trimdata/{sample}{read}"+ext
    output:
        "fastQC/{sample}{read}_fastqc.html"
    log:
        "fastQC/logs/{sample}{read}.log"
    shell:
        "fastqc -o fastQC -t {threads} -q {input} 2> {log}"

rule hisat2:
    input:
        r1="trimdata/{sample}"+reads[0]+ext,
        r2="trimdata/{sample}"+reads[1]+ext
    output:
        bam="alignment/{sample}.bam",
        summary="alignment/{sample}.summary"
    threads: 24
    params:
        index=genome_index,
        splice="alignment/{sample}.splice.txt",
        met="alignment/{sample}.matrix.txt"
    log:
        "alignment/logs/{sample}.align.log"
    shell:
        "hisat2 -q --rna-strandness RF --dta-cufflinks -x {params.index} -1 {input.r1} -2 {input.r2} -p {threads} --novel-splicesite-outfile {params.splice} --summary-file {output.summary} --met-file {params.met} | "
        "samtools view -Sbh - > {output.bam} 2> {log}"
#é“¾ç‰¹å¼‚æ€§ï¼Œå¦‚æœä½¿ç”¨dUTP: --rna-strandness RFï¼Œé»˜è®¤éé“¾ç‰¹å¼‚æ€§

rule filter:
    input:
        "alignment/{sample}.bam"
    output:
        "alignment/{sample}.filter.sorted.bam"
    threads: 16
    log:
        "alignment/logs/{sample}.filter.log"
    shell:
        "samtools view -b -F 780 -f 2 -q 30 -@ {threads} {input} | "
        "samtools sort -@ {threads} -o {output} 2> {log}"

rule bamindex:
    input:
        "alignment/{sample}.filter.sorted.bam"
    output:
        "alignment/{sample}.filter.sorted.bam.bai"
    shell:
        "samtools index {input}"

rule stringtie:
    input:
        "alignment/{sample}.filter.sorted.bam"
    output:
        gtf="counts/{sample}/{sample}.gtf",
        tab="counts/{sample}.tab"
    params:
        ann=genome_gtf
    log:
        "counts/logs/{sample}.log"
    shell:
        "stringtie -o {output.gtf} -A {output.tab} -e -G {params.ann} --rf {input} 2> {log}"

rule featurecounts:
    input:
        "alignment/{sample}.filter.sorted.bam"
    output:
        "featurecounts/{sample}.count.txt"
    params:
        gtf=genome_gtf
    log:
        "featurecounts/logs/{sample}.count.log"
    shell:
        "featureCounts -s 2 -p -B -t exon -g gene_id -a {params.gtf} -o {output} {input} 2> {log}"   
```

### å·®å¼‚è¡¨è¾¾åˆ†æ

ä½¿ç”¨RåŒ…DESeq2è¿›è¡Œå·®å¼‚åˆ†æï¼Œ
```
library(tidyverse)

setwd('featurecounts')

#åˆå¹¶featurecountsç»“æœ
files <- list.files(pattern = '.count.txt$')
files
counts_d <- NULL
for (i in files){
  d <- read_tsv(i,skip = 1)
  n <- str_sub(i,end=-11)
  n2 <- paste0(n,'_fpkm')
  d <- select(d,c(1,6,7))
  colnames(d) <- c('geneid','length',n)
  N <- sum(d[,n])
  d[,n2] <- round(10^9*d[,n]/(d[,'length']*N),2)
  # d <- mutate(d,fpkm=round(10^9*d[,n]/(d[,'length']*N),2))#è®¡ç®—fpkm
  # colnames(d) <- c('geneid','length',n,n2)
  #colnames æ— æ³•æ”¹åï¼Ÿï¼Ÿï¼Œ
  if (length(counts_d)==0){
    counts_d <- d
  }else{
    counts_d <- merge(counts_d,d,by=c('geneid','length'))
  }
}
head(counts_d)
write.table(counts_d,'expr_of_rnaseq.txt',sep='\t',row.names = F)

#ç›¸å…³æ€§åˆ†æ
corr_d <- counts_d[,c(4,6,8,10)]  #å–fpkmå€¼åšç›¸å…³æ€§åˆ†æï¼Œæ³¨æ„æ ·æœ¬æ•°é‡
x <- rowSums(corr_d)
corr_d <- corr_d[x>0,]

# library(ggcorrplot)

# corr <- round(cor(log2(corr_d+1)),2)
# ggcorrplot(
#   corr,
#   hc.order = TRUE,
#   # type = "lower",
#   outline.color = "white",
#   ggtheme = ggplot2::theme_gray,
#   colors = c("#6D9EC1", "white", "#E46726"),
#   lab=TRUE
# )
# ggsave('corrplot.jpg',width = 14,height = 14,dpi = 300,units = 'cm')

library(GGally)

ggpairs(log2(corr_d+1),columnLabels=c('ctrl_#1','treat_#1','ctrl_#2','treat_#2'))+theme_bw()
ggsave('pairsplot.jpg',width = 14,height = 14,dpi = 300,units = 'cm')

## åŸºå› å·®å¼‚è¡¨è¾¾åˆ†æ
library(DESeq2)
cts <- counts_d%>%select(1,3,5,7,9) # åŸºå› è¡¨è¾¾çŸ©é˜µï¼ŒåŸºå› åä½œä¸ºè¡Œåï¼Œæ ·æœ¬åä½œä¸ºåˆ—åï¼Œè‡³å°‘æœ‰2ä¸ªé‡å¤
rownames(cts) <- cts$geneid
cts$geneid <- NULL
design <- data.frame(
  sample = colnames(cts),
  condition = c("ctrl", "treat", "ctrl", "treat"),
  batch = c("I", "I", "II", "II")
) #è®¾ç½®æ‰¹æ¬¡ï¼Œå»é™¤æ‰¹æ¬¡æ•ˆåº”
coldata <- data.frame(row.names = design$sample, condition = factor(design$condition), batch = factor(design$batch))
dds <- DESeqDataSetFromMatrix(cts, colData = coldata, design = ~ batch + condition) # å»é™¤æ‰¹æ¬¡æ•ˆåº”~batch+condition
dds$condition <- factor(dds$condition, levels = c("treat", "ctrl")) # è®¾ç½®æ¯”è¾ƒæ–¹å‘ï¼Œå¦‚treat-ctrl
dds <- DESeq(dds)

# PCA åˆ†æåŠå¯è§†åŒ–
vsd <- vst(dds, blind = FALSE)
plotPCA(vsd, "condition")
assay(vsd) <- limma::removeBatchEffect(assay(vsd), vsd$batch)
plotPCA(vsd, "condition")  #è§‚å¯Ÿå»é™¤æ‰¹æ¬¡æ•ˆåº”åçš„æ•ˆæœ
# pcaData <- plotPCA(vsd, ntop = 2000, intgroup = "condition", returnData = TRUE)
# percentVar <- round(100 * attr(pcaData, "percentVar"))
# pcaData %>%
#   ggplot(aes(PC1, PC2, color = condition)) +
#   geom_point(size = 5) +
#   geom_text(aes(label = name)) +
#   scale_color_manual(values = c("darkred", "darkblue")) +
#   xlab(paste0("PC1: ", percentVar[1], "% variance")) +
#   ylab(paste0("PC2: ", percentVar[2], "% variance")) +
#   theme_base()

# å·®å¼‚è¡¨è¾¾åˆ†æ
res <- results(dds, contrast = c("condition", "treat", "ctrl"))
res <- res[order(res$pvalue), ]
res <- na.omit(as.data.frame(res))
res$geneid <- rownames(res)

# å®šä¹‰æ˜¾è‘—æ€§é˜ˆå€¼ï¼Œç­›é€‰å·®å¼‚è¡¨è¾¾åˆ†æ
p <- log10(0.05) * (-1)
fc <- log2(2)
degs <- res %>%
  select(geneid, log2FoldChange, pvalue, padj) %>%
  mutate(logp = log10(pvalue) * (-1)) %>%
  mutate(type = if_else(logp > p, if_else(log2FoldChange > fc, "Up", if_else(log2FoldChange < (-fc), "Down", "N.s")), "N.s"))
table(degs$type)
write.table(degs, "degs_of_rnaseq.txt", row.names = F, sep='\t')

# ç«å±±å›¾å¯è§†åŒ–
degs %>%
  filter(log2FoldChange < 10) %>%
  filter(logp < 50) %>%
  ggplot(aes(log2FoldChange, logp)) +
  geom_point(aes(colour = type), alpha = 0.8, size = 0.8) +
  scale_colour_manual(values = c("#3C5488FF", "grey", "#DC0000FF")) +
  guides(colour = guide_legend(override.aes = list(alpha = 1))) +
  # geom_vline(xintercept = c(-fc, fc), linetype = 2) +  #æ·»åŠ æŒ‡ç¤ºçº¿
  # geom_hline(yintercept = p * (-1), linetype = 2) +
  scale_x_continuous(breaks = seq(-4, 4, 2), labels = seq(-4, 4, 2), limits = c(-5, 5)) +
  labs(x = quote(log[2] ~ FoldChange), y = quote(-log[10] ~ pvalue), colour = "") +
  theme_light() +
  theme(
    panel.border = element_rect(colour = "black"),
    axis.ticks = element_line(colour = "black"),
    axis.text = element_text(colour = "black", size = 14),
    axis.title = element_text(size = 14),
    legend.key.height = unit(0.4, "cm"), legend.key.width = unit(0.4, "cm")
  )
ggsave("vocalno.jpg", width = 10, height = 8, dpi = 1200, units = "cm")

```

### åŠŸèƒ½å¯Œé›†åˆ†æ

---
ç¼…æ€€æ¯›ä¸»å¸­é€ä¸–46å‘¨å¹´Â·1976å¹´9æœˆ9æ—¥è‡³2022å¹´9æœˆ9æ—¥ ğŸ’®ğŸ•¯ğŸ•¯ğŸ•¯
